tools/validator.py

```
@@ -0,0 +1,152 @@
+            result['passed'] = False
+
+        logger.info(f"Security validation: {len(result['critical_issues'])} critical, "
+                   f"{len(result['high_issues'])} high, {len(result['medium_issues'])} medium")
+
+        return result
+
+    async def _run_semgrep(self, files: Dict) -> Optional[Dict]:
+        """Run semgrep if available"""
+        try:
+            # Check if semgrep is installed
+            result = subprocess.run(
```

Ruff (0.14.10)

Async functions should not run processes with blocking methods

(ASYNC221)